### PR TITLE
SMC fix

### PR DESCRIPTION
**Pull request by lintusj1**
Reimplementation of SMC

This is a preliminary PR that may require some more polishing still.

#### Summary:

- SMC is now able to handle any DAG structure
- SMC works now asynchronously
- added additional functionality to core to support the new implementation
- Small improvements to Rejection sampling

### CODE REVIEW

File: elfi/methods.py [hunk 1]
```diff
@@ -239,60 +291,121 @@ def sample(self, n_samples, n_populations, schedule):
...
+
+            while True:
+                # In the first round use accept rate of the previous round
+                self._add_new_batches(n_samples, n_accepts_t, accept_rate)
+
+                all_new_distances, i_future, remaining_futures = async.wait(self.distance_futures)
+                self.distance_futures = remaining_futures
+                p_outs = self.parameter_futures.pop(i_future)
+
+                new_accepts = all_new_distances[:,0] <= threshold
```
Comment (vuolleko) [inline elfi/methods.py]: Use _compute_acceptance()?

File: elfi/utils.py [hunk 2]
```diff
@@ -129,18 +183,3 @@ def stochastic_optimization(fun, bounds, its, polish=False):
                                     polish=polish, init='latinhypercube')
     return result.x, result.fun
 
-
-def weighted_var(data, weights):
```
Comment (vuolleko) [inline elfi/utils.py]: Why remove this generally useful function?

File: elfi/distributions.py [hunk 3]
```diff
@@ -222,43 +230,106 @@ def __init__(self, *args, observed=None, size=None, **kwargs):
...
+            xi = x[i,:] - self._samples
+            vals[i] = np.sum(self.p_weights * d.pdf(xi))
+        return vals
+
+    @property
+    def size(self):
+        return self._samples[0].shape
+
+    @property
+    def weighted_cov(self):
```
Comment (vuolleko) [inline elfi/distributions.py]: Why define a general function inside property?

File: elfi/distributions.py [hunk 4]
```diff
@@ -222,43 +230,106 @@ def __init__(self, *args, observed=None, size=None, **kwargs):
...
+        """Unbiased weighted covariance"""
+        x = self._samples.copy()
+        w = self.p_weights
+        x -= np.average(x, axis=0, weights=w)
+
+        a = 1/(1-np.sum(w**2))
+        if np.isinf(a):
```
Comment (vuolleko) [inline elfi/distributions.py]: Use try instead?

File: elfi/core.py [hunk 5]
```diff
@@ -566,12 +546,13 @@ def _inherit_observed(self):
         return observed
 
 
-"""
-Operation nodes
-"""
+def constant_transform(input_dict, constant):
```
Comment (lintusj1) [inline elfi/core.py]: Added a comment.

File: elfi/core.py [hunk 6]
```diff
@@ -588,16 +569,69 @@ def __init__(self, name, value):
...
+        transform : callable
+            transform build from the operation and kwargs
+        kwargs : dict
+            remaining kwargs not used in making the transform
+        """
+        self._operation = operation
+        transform = partial(self.__class__.operation_transform, operation=operation)
+        return transform, kwargs
+
+    def redefine(self, operation, *parents, reset=True, **kwargs):
```
Comment (lintusj1) [inline elfi/core.py]: Added

File: elfi/core.py [hunk 7]
```diff
@@ -637,11 +671,11 @@ class Simulator(ObservedMixin, RandomStateMixin, Operation):
     name: string
     simulator: callable
     """
-    operation_wrapper = simulator_wrapper
+    operation_transform = simulator_transform
 
 
 # TODO: rename to standard_wrapper
-def summary_wrapper(operation, input_dict):
+def summary_transform(input_dict, operation):
```
Comment (lintusj1) [inline elfi/core.py]: Added

File: elfi/core.py [hunk 8]
```diff
@@ -656,10 +690,10 @@ def summary_wrapper(operation, input_dict):
 
 
 class Summary(ObservedMixin, Operation):
-    operation_wrapper = summary_wrapper
+    operation_transform = summary_transform
 
 
-def discrepancy_wrapper(operation, input_dict):
+def discrepancy_transform(input_dict, operation):
```
Comment (lintusj1) [inline elfi/core.py]: Added

File: elfi/methods.py [hunk 9]
```diff
@@ -76,28 +76,57 @@ def sample(self, n_samples, *args, **kwargs):
         """
         raise NotImplementedError("Subclass implements")
 
-    def _get_distances(self, n_samples):
-        """Run the all-accepting sampler.
+    def estimate_batches_needed(self, n_samples, accept_rate):
```
Comment (lintusj1) [inline elfi/methods.py]: Added

File: elfi/methods.py [hunk 10]
```diff
@@ -76,28 +76,57 @@ def sample(self, n_samples, *args, **kwargs):
...
+        return self.estimate_proposals_needed(n_samples, accept_rate) // self.batch_size
+
+    def estimate_proposals_needed(self, n_samples, accept_rate):
+        """The returned estimate is divisible with self.batch_size"""
```
Comment (lintusj1) [inline elfi/methods.py]: Added additional explanation. What is by the way the minimal format for a docstring?

File: elfi/methods.py [hunk 11]
```diff
@@ -76,28 +76,57 @@ def sample(self, n_samples, *args, **kwargs):
...
+            logger.info("{}: Running with {} proposals.".format(self.__class__.__name__,
+                                                                n_sim))
+        distances = self.distance_node.acquire(n_sim,
+                                               batch_size=self.batch_size).compute()
+        parameters = [p.acquire(n_sim, batch_size=self.batch_size).compute()
                       for p in self.parameter_nodes]
 
         return distances, parameters
 
+    def _ceil_in_batch_sizes(self, n):
```
Comment (lintusj1) [inline elfi/methods.py]: Added.

File: elfi/methods.py [hunk 12]
```diff
@@ -108,9 +137,7 @@ def sample(self, n_samples, quantile=0.01, threshold=None):
         In quantile mode, the simulator is run (n/quantile) times.
 
         In threshold mode, the simulator is run until n_samples can be returned.
-        DANGER: a poorly-chosen threshold may result in a never-ending loop.
-
-        TODO: handle cases with vector discrepancy
+        Not that a poorly-chosen threshold may result in a never-ending loop.
```
Comment (lintusj1) [inline elfi/methods.py]: Fixed.

File: elfi/methods.py [hunk 13]
```diff
@@ -135,34 +162,39 @@ def sample(self, n_samples, quantile=0.01, threshold=None):
...
-                accepted = distances < threshold
-                n_accepted = sum(accepted)
+        parameters = None; distances = None; accepted = None; accept_rate = None
+
+        if threshold is None:
+            # Quantile case
+            n_sim = int(np.ceil(n_samples / quantile))
+            distances, parameters = self._acquire(n_sim)
+            sorted_inds = np.argsort(distances, axis=0)
+            threshold = distances[sorted_inds[n_samples-1]].item()
```
Comment (lintusj1) [inline elfi/methods.py]: I'm not sure if I understood. This get's the distance of the last accepted sample in order.

File: elfi/methods.py [hunk 14]
```diff
@@ -190,16 +222,31 @@ def reject(self, threshold, n_sim=None):
         if n_sim is None:
             n_sim = self.distance_node._generate_index
 
-        distances, parameters = self._get_distances(n_sim)
-        distances = distances.ravel()  # avoid unnecessary indexing
+        distances, parameters = self._acquire(n_sim)
+        accepted = distances[:, 0] < threshold
```
Comment (lintusj1) [inline elfi/methods.py]: Fixed

File: elfi/methods.py [hunk 15]
```diff
@@ -190,16 +222,31 @@ def reject(self, threshold, n_sim=None):
...
+                    threshold=threshold,
+                    n_sim=n_sim,
+                    accept_rate=accept_rate)
 
-        accepted = distances < threshold
-        posteriors = [p[accepted] for p in parameters]
 
-        return {'samples': posteriors, 'threshold': threshold, 'n_sim': n_sim}
+# TODO: add tests
+def smc_prior_transform(input_dict, column_interval, prior):
```
Comment (lintusj1) [inline elfi/methods.py]: Added a docstring. Will have to think of a better place.

File: elfi/methods.py [hunk 16]
```diff
@@ -216,6 +263,11 @@ class SMC(Rejection):
     `Rejection` : Basic rejection sampling.
```
Comment (lintusj1) [inline elfi/methods.py]: Updated

File: elfi/utils.py (hunk 2 above)
Comment (lintusj1) [inline elfi/utils.py]: It is now replaced with `weighted_cov`.

File: elfi/distributions.py (hunk 3 above)
Comment (lintusj1) [inline elfi/distributions.py]: Moved to `utils.py`.

File: elfi/distributions.py (hunk 4 above)
Comment (lintusj1) [inline elfi/distributions.py]: Refactored.

**Pull request merged**